src/sagemaker/serve/utils/hardware_detector.py

```diff
@@ -0,0 +1,107 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Utilites for identifying and analyzing instance gpu hardware"""
+from __future__ import absolute_import
+
+import logging
+
+from botocore.exceptions import ClientError
+
+from sagemaker import Session
+
+logger = logging.getLogger(__name__)
+
+FALLBACK_GPU_RESOURCE_MAPPING = {
+    "ml.p5.48xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 655360},
+    "ml.p4d.24xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 327680},
+    "ml.p4de.24xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 655360},
+    "ml.p3.2xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.p3.8xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 65536},
+    "ml.p3.16xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 131072},
+    "ml.p3dn.24xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 262144},
+    "ml.p2.xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 12288},
+    "ml.p2.8xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 98304},
+    "ml.p2.16xlarge": {"Count": 16, "TotalGpuMemoryInMiB": 196608},
+    "ml.g4dn.xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.2xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.4xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.8xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.16xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.12xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 65536},
+    "ml.g5n.xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.2xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.4xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.8xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.16xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.12xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 98304},
+    "ml.g5.24xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 98304},
+    "ml.g5.48xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 196608},
```

Comment:
Can you detail the process in place so when a new accelerator hardware gets released in SageMaker, how we automate the logic, so the end to end release of an instance is taken care in PySDK as well?

I understand it is a fall back logic in case of access denied. But, I am curious on how we plan to keep it fresh.

Reply:
You are right about the static list being out of date, the plan long term is to have a table (DynamoDB or otherwise) that has a lambda/wf to fetch additional fallback types in an instance family or at the minimum the ability to append new types. 

Proposed process we are going for:
1. Customer provides new accelerator hardware , we query Ec2 API to get details. 
2. If unable to query API, fallback to this static list. 
3. Try HuggingFace Hub with additional parameters
4. Fail and propagate

Reply:
As of now, we do not have an automatic way to maintain the collection. Will discuss with the team on how we plan to refresh this moving forward.